### ISSUE TITLE
Popover: focusOnMount shouldn't be specific to keyboard interaction

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#5559]
Many users are not exclusively "mouse users" or "keyboard users". Actually, many users make a mixed use of mouse and keyboard. Implementing software features that work just with mouse or keyboard interaction, is often just an assumption.

In the Popover component [there's some code](https://github.com/WordPress/gutenberg/blob/8a4d644a525c1463f469e525d5d31ad36f44e23c/components/popover/index.js#L22) intentionally used to move focus to the Popover only when a keyboard is used.

This seems to me a big assumption, as focus should always be moved to the Popover regardless of the device used. Anecdotally, I personally often make a mixed use of mouse and keyboard, as that's the fastest way to interact with an interface for me. I'm pretty sure I'm not the only one.

So what happens with the Popover? Let's consider one of the Popover implementations: the more options menu. When using the mouse (screenshot on the left) focus is _not_ moved to the Popover. Instead, when using a keyboard (screenshot on the right) focus is moved to the Popover:

<img width="650" alt="mouse and keyboard" src="https://user-images.githubusercontent.com/1682452/37258236-12182e0a-2575-11e8-8c4c-238ae1e89db3.png">

This doesn't allow mixed use of mouse and keyboard: for example, users might want to click to open the menu and then quickly use the Tab key to select a menu item. Users often find unexpected ways to interact with an UI, an the UI shouldn't make any assumption on the expected behaviors.

I don't see a good reason for this distinction between mouse and keyboard interaction other than the will to not display the focus style on the first item. If this is the case, I'd like to remind everyone the focus style is a native accessibility feature that should always work for all users, regardless of the device they use.

From a technical perspective, removing this different behavior would also help in reducing the number of keyboard events used in Gutenberg, which have recently come to a point where they seem [rather convoluted and fragile](https://github.com/WordPress/gutenberg/issues/5501#issuecomment-371949740). /cc @aduth

**Comment by aduth** [comment]
>focus should always be moved to the Popover regardless of the device used

In retrospect, I can agree with this point. The problem I'd encountered, as demonstrated in the "Before" screenshots of #5318, is that the styling applied in shifting focus to the first item can confuse intention between states of selection and focus.

![focus](https://user-images.githubusercontent.com/1779930/36815915-3f623438-1caa-11e8-90e4-18607ed17ecf.png)

Perhaps to address this (and arguably for better consistency with other similar native UIs), could it be that opening the menu would still shift focus to the popover, but to its container, not the first item? Then the user could still proceed to tab/arrow would navigate options in the list.

![menu](https://user-images.githubusercontent.com/1779930/37263741-e367470c-257f-11e8-9cbc-19f8a3c44d9a.gif)

In the meantime, I would be fine to revert #5318, and can plan to take a look at an alternate implementation this week.

**Comment by afercia** [comment]
> for better consistency with other similar native UIs)

The Finder example works a bit differently when using the keyboard:
- Ctrl + F5 to focus the finder window menu bar
- Tab to move to the "Arrange" button
- press spacebar or arrow down to open the menu
- see below: focus stays on the Arrange button (notice the outline)

<img width="311" alt="screen shot 2018-03-12 at 09 11 15" src="https://user-images.githubusercontent.com/1682452/37272307-8f281174-25d6-11e8-8738-3ba6d075a825.png">

Of course, finder is a desktop application and works differently from a web application, but this behavior is more similar to what we've described a few times for toggles and panels: ideally, the expanded panel (or popup, whatever we want to call it) should immediately follow the control that opened it. This way, it is just placed in the natural tab order.

Instead, when it's placed far in the source, there's the need to manage focus. I can only refer to the expected behaviors listed in the ARIA authoring practices for similar UIs and they all ask to set initial focus on an element _inside_ the panel/menu/dialog:

https://www.w3.org/TR/wai-aria-practices/#menubutton
> opens the menu and places focus on the first menu item.

https://www.w3.org/TR/wai-aria-practices/#dialog_modal
> When a dialog opens, focus moves to an element inside the dialog. See notes below ...

The issue in #5318 seems to me a styling issue, can we try to solve it on the styling side, trying to clearly differentiate the "fcused" and "active" styles?

**Comment by aduth** [comment]
>>https://www.w3.org/TR/wai-aria-practices/#menubutton
>
>opens the menu and places focus on the first menu item.

The full context of the quote is in pressing Enter/Space/Up/Down while the button is already focused:

>With focus on the button:
>
> - Enter: opens the menu and places focus on the first menu item.
> - Space: Opens the menu and places focus on the first menu item.
> - (Optional) Down Arrow: opens the menu and moves focus to the first menu item.
> - (Optional) Up Arrow: opens the menu and moves focus to the last menu item.

Assuming the act of clicking on the button as its focusing, would it then not stand to reason the first menu item would only become focused after then proceeding to press one of these keys?

**Comment by afercia** [comment]
To be honest, _mouse_ interaction is not described in the ARIA authoring practices. I see their examples (there are a few linked on the menu and menubar sections) work differently with mouse; they all reveal the menu on hover 😬 Just _some_ of them work on mouse click too and move focus to the first menu item:
https://www.w3.org/TR/wai-aria-practices/examples/menu-button/menu-button-actions.html
https://www.w3.org/TR/wai-aria-practices/examples/menubar/menubar-2/menubar-2.html

I'd rather try to keep the functionality as simple as possible, without assumptions. Styling issues should be solved on the presentation side.

**Comment by aduth** [comment]
The simplest approach would be to assume that the button should first become focused, then navigation to menu item occurs when proceeding to press one of the above-mentioned buttons.

**Comment by aduth** [comment]
>Technically, the button receives focus while it's clicked.

Precisely, so the first step is satisfied. The user can then press [some aforementioned button] to navigate to the first item in the list.

I intentionally avoided mention of interaction device in my previous comment to align behaviors on "in focus, and out of focus".

**Comment by aduth** [comment]
Related: #5595
